app/slave/cluster_slave.py

```diff
@@ -74,24 +75,36 @@ def _start_heartbeat(self):
 
     def _run_heartbeat(self):
         try:
-            self._send_heartbeat_to_master()
-            self._heartbeat_failure_count = 0
+            response = self._send_heartbeat_to_master()
+
+            # A non HTTP 200 response indicates that the master is up, but does not recognize the slave.
+            # The slave tries to reconnect to master at this point.
+            # This is also counted as a heartbeat failure and the slave takes an action accordingly
+            if response.status_code != http.client.OK:
+                self._act_on_heartbeat_failure()
+                self._logger.info('Master responded with {}. Attempting to reconnect.'.format(response.status_code))
+                self.connect_to_master(self._master_url)
```

Comment:
1. This connect_to_master method adds 2 callbacks that will trigger when someone kills the slave. If this gets called a lot of times I think multiple instances of those callbacks will get collected and called multiple times.

2. I don't think we should be calling connect_to_master at all here. I talked with @cmcginty offline about this and it seems like there might've been some confusion. My understanding was always that there would be no difference between sending a heartbeat and sending a request to connect to the master. The slave would periodically send a heartbeat and if the master didn't know about that slave it would connect it, if the master did know about that slave then it would be a noop. So you never have to explicitly connect to the master, the heartbeat request takes care of that for you